### PR TITLE
feat: trial table [DET-3019]

### PR DESCRIPTION
**Pull request by hkang1**
## Description

UPDATED:
Screen Shot 2020-08-17 at 3 35 54 PM

## Commentary (optional)

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: webui/react/src/utils/trial.ts [hunk 1]
```diff
@@ -1,5 +1,13 @@
 import { MetricNames, Step } from 'types';
-import { isNumber } from 'utils/data';
+import { alphanumericSorter, isNumber } from 'utils/data';
+
+export const extractMetricValue = (step: Step, metricName: string): number | undefined => {
```
[the remark this answers is not in the document]
Comment (hkang1) [reply webui/react/src/utils/trial.ts:4]: great call, did not think about the same name being used for both types of metric!

File: webui/react/src/pages/TrialDetails.tsx [hunk 2]
```diff
@@ -1,44 +1,99 @@
...
+          ...record.checkpoint,
+          batch: record.numBatches + record.priorBatchesProcessed,
+          experimentId: trial.data?.experimentId,
+          trialId: record.id,
+        };
+        return <Tooltip title="View Checkpoint">
+          <Button
+            aria-label="View Checkpoint"
+            icon={<Icon name="checkpoint" />}
+            onClick={e => handleCheckpointShow(e, checkpoint)} />
```
[the remark this answers is not in the document]
Comment (hkang1) [reply webui/react/src/pages/TrialDetails.tsx:84]: This is a great topic to talk about during WebUI-AG. Thinking something like this for the general component content order?
```
const Component = (props) => {
  ...context api hooks (useStateContext, useActionContext)
  ...useStorage
  ...useStates
  ...useRestApis

  ...useMemos

  ...useEffects

  ...useLayoutEffects

  ...useCallbacks

  if (...) return JSX;
  else if (...) return JSX;
  return (JSX);
};
```

[the remark this answers is not in the document]
Comment (hamidzr) [reply webui/react/src/pages/TrialDetails.tsx:84]: yeah I agree some order is good. although maybe we'd keep component logic small enough by creating more component then it wouldn't be as important but that's more work

**Comment by hkang1** [comment]
> This is looking great and everything works. As we talked there might be an issue with duplicate metric names between training and validation and some of the logic. I added some comments.
> 
> could we either limit the precision of these numbers (accompanying a tooltip) or use the humanizedNumber helper? I'm not sure what AMLE would prefer.
> 

Talked with some AMLE folks and it seems like scientific notation is commonplace. Updated to use the `humanReadableFloat` (scientific notation) plus a tooltip to show the original value per your suggestion.

> * it seems to be missing the has checkpoint filter
> 

Oops forgot this filter, it should be in now.

File: webui/react/src/utils/data.ts [hunk 3]
```diff
@@ -90,6 +90,21 @@ export const commandStateSorter = (a: CommandState, b: CommandState): number =>
   return commandStateSortValues[a] - commandStateSortValues[b];
 };
 
+/*
+ * Sort the metric names by having the validation metrics come first followed by training metrics.
+ * Within each type of metric, sort in the order they appear in the `MetricNames` array.
+ * Within the respective type of metrics, `MetricNames` is currently sorted alphanumerically.
+ */
+export const metricNameSorter = (metricNames: MetricNames) => (a: string, b: string): number => {
```
[the remark this answers is not in the document]
Comment (hkang1) [reply webui/react/src/utils/data.ts]: The changes of having a `MetricType` (`Validation` or `Training`) should now be able to distinguish metrics with the same name.

File: webui/react/src/components/SelectFilter.tsx [hunk 4]
```diff
@@ -7,8 +7,11 @@ import Label from './Label';
 import css from './SelectFilter.module.scss';
 
 interface Props {
+  defaultValue?: SelectValue;
   enableSearchFilter?: boolean;
   label: string;
+  mode?: 'multiple' | 'tags';
```
[the remark this answers is not in the document]
Comment (hkang1) [reply webui/react/src/components/SelectFilter.tsx:13]: I was hoping to find something like what we had in Elm, but this one is lacking in `antd`. We would have to customize this ourselves.

**Comment by hkang1** [comment]
> looks great!
> question: wdy think about using a smaller label for showing validation vs training type? maybe `t` or `v` with a tooltip? it should be something we'd use a lot in the UI. if we think a smaller label or icon could be as informative to the user with minimal training I think that makes the view cleaner

Love the idea of the smaller tag with a tooltip!

File: webui/react/src/components/SelectFilter.tsx [hunk 5]
```diff
@@ -1,29 +1,28 @@
...
-import { SelectValue } from 'antd/es/select';
+import { SelectProps, SelectValue } from 'antd/es/select';
 import React, { PropsWithChildren, useCallback } from 'react';
 
 import Icon from './Icon';
 import Label from './Label';
 import css from './SelectFilter.module.scss';
 
-interface Props {
+interface Props<T = SelectValue> extends SelectProps<T> {
```
Comment (hamidzr) [inline webui/react/src/components/SelectFilter.tsx:9]: :+1:

**Comment by hamidzr** [comment]
> Love the idea of the smaller tag with a tooltip!

: ) I'm thinking we might use color os use a bit lighter background for the label later

**Pull request merged**